### ISSUE TITLE
SAR images with complex data need to have non-complex bands

### ISSUE THREAD

**Issue by mortenwh** [nansencenter/nansat#39]
Many packages depending on nansat expect some standard SAR bands like sigma0 in VV pol to be in real numbers. It will therefore be useful to adjust the SAR mappers to use the IntensityPixelFunc to provide sigma0 in magnitude with band names sigma0_VV, sigma0_HH, etc. I propose to add "_complex" to the suffix of bands with complex data.

**Comment by asumak** [comment]
If 'raw_counts' band is complex, it is necessary to set 'SourceTransferType' in VRT. If 'raw_counts'  and 'incidence angle' bands are passed to the pixel functions, 'SourceTransferType' is applied to the both bands.It is impossible to set 'SourceTransferType'  to each band. Therefore the extra bands, converted from complex to real and imaginary parts, have to be added.

**Comment by asumak** [comment]
Can you please check bb4bdbe, Morten? 
And please correct wkt, sphPass, other metadata. (line 204, 205, 238 etc...)

**Comment by mortenwh** [comment]
There is a lot of code changes in bb4bdbed9644d81e8c8b99712d2744221fea7f29 - have you written a test?
Other than that, export function now fails when there is complex data. Could you check that as well?

**Comment by asumak** [comment]
OK. I will do that.

**Comment by asumak** [comment]
When do you get the fails with complex data, Morten? 
It is not always, is it?

**Comment by mortenwh** [comment]
I tried to export a nansat object which has complex bands - could you try
that too?

**Comment by asumak** [comment]
yes, I used 'ASA_IMS_1PNIPA20100411_101715_000000162088_00280_42418_0338.N1'.
the 1st and 7th bands are complex. the 7th is created by pixelfunctions.

n=Nansat('ASA_IMS_1PNIPA20100411_101715_000000162088_00280_42418_0338.N1)
---------------------
n.vrt.delete_bands([1,2,3,4,5,6]) or n.vrt.delete_bands([2,3,4,5,6,7])
n.export('ofile')
worked fine.
--------------------
n1 = Nansat(domain=Domain(ds=n.vrt.dataset), array=n[7])
n1.export('ofile')
worked fine.
--------------------
n.export('ofile', bands=[7])
failed.

**Comment by mortenwh** [comment]
I'm just asking this: 
```
n=Nansat('ASA_IMS_1PNIPA20100411_101715_000000162088_00280_42418_0338.N1)
n.export('outfile.nc')
```
It fails, right? Can you make a separate issue about this, if it is the case? I can probably delete the complex bands as a temporary workaround.

**Comment by mortenwh** [comment]
I checked the code above myself now, and the example seems to work (although tested with RS2). My problem must be something else. I get this error message: 
```
*** RuntimeError: RuntimeError('netcdf error #-62 : NetCDF: One or more variable sizes violate format constraints .\nat (netcdfdataset.cpp,~netCDFDataset,1551)\n',)
```
I assumed it was because of the complex data but now I'm not sure... If I delete the complex bands like you did in your example above, it works. What about your last example, shouldn't that work as well?

**Comment by mortenwh** [comment]
btw - now I noticed that the export function duplicates the last band in my nansat object (I have a last band called "noise"). After deleting the complex bands and exporting, I have another band called "noise_001". We should find out of this as well, although it's not crucial....

**Comment by knutfrode** [comment]
The error message "NetCDF: One or more variable sizes violate format constraints" normally means that one tries to create a file larger than permitted for the file format, normally 2 GB for netCDF.
http://www.unidata.ucar.edu/software/netcdf/faq-lfs.html

**Comment by mortenwh** [comment]
Aha - then that explains everything :) thanks knutfrode! So, I guess it remains to make a good test for this issue, and perhaps do a quick check of the added band at export..

**Comment by asumak** [comment]
Although I tested export(), I do not know when a duplicate band is created. 
Can you tell me the detail, Morten?

**Comment by mortenwh** [comment]
You're right, it doesn't happen in the simple example. I'm rerunning my processing now to check once more

**Comment by mortenwh** [comment]
Nevermind - it was a bug in a function of a class inheriting from nansat. Then it is good :)

**Comment by asumak** [comment]
I made a test for ASAR sigma0 band in 5cf64e6.
Although I can get complex data array, I get an error as " Error in `python': munmap_chunk(): invalid pointer: 0x0a17c6e8".
It seems a memory problem of my virtual machine.
Can you test it, Morten?

**Comment by mortenwh** [comment]
I have rewritten the test to prepare it for testing all SAR mappers. I also changed the `standard_name` of complex data in the asar mapper. The standard_name of the Normalized Radar Cross Section should be "surface_backwards_scattering_coefficient_of_radar_wave" regardless of whether it is complex or not. By using the standard_name, it is not so important if mappers don't follow the name-convention suggested above. In order to get the test returning all complex bands (not only one) I had to change the function `Nansat._get_band_number`. Please double-check that this is ok. Can you extend the test to also take Sentinel1 and others?

The test is failing on the ASAR example file `ASA_IMS_1PNIPA20100411_101715_000000162088_00280_42418_0338.N1` at line 171 with "GDALError: Cannot read array from band 7", so there must be a bug somewhere.

**Comment by mortenwh** [comment]
See 38b17120dd9d5b5a4c84b534eed142f628fa5200 for latest commit

**Comment by asumak** [comment]
Regarding "GDALError: Cannot read array from band 7":
Isn't it a memory problem? 
I have the same problem. But after resizeing with eResampleAlg=0, I can read the array although I get " Error in `python': munmap_chunk(): invalid pointer:.

**Comment by mortenwh** [comment]
I had forgotten to compile the pixel functions... The ASAR file now passes the test, but there is a radarsat-2 product that doesn't - can you check @asumak?

**Comment by asumak** [comment]
ok. I do.

**Comment by asumak** [comment]
The problem with RS2 was solved in a508744. 
I will continue for setinel1.

**Comment by asumak** [comment]
Added 'name' for getting a band number in 6fee462 because some seninel-1 data have some bands whose names are 'DN_XX'. 
e.g. 'S1A_IW_SLC__1SDV_20150317T163128_20150317T163156_005074_0065FB_2245.zip' has DN_VH, DN_VH_000 and DN_VH_001.

**Comment by asumak** [comment]
I think it is good to make tests for sentinel-1 when we modify the mapper for complex data.
